docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1479 @@
+
+    def _solve_internal_problem(
+        self, problem: InternalProblem, x0: NDArray[float]
+    ) -> InternalOptimizeResult:
+
+        options = {
+            "maxiter": self.stopping_max_iterations,
+            "maxfev": self.stopping_max_criterion_evaluations,
+            "xatol": self.convergence_absolute_params_tolerance,
+            "fatol": self.convergence_absolute_criterion_tolerance,
+            "adaptive": self.adaptive,
+        }
+
+        res = minimize(
+            fun=problem.scalar.fun,
+            x0=x,
+            bounds=_get_scipy_bounds(problem.bounds),
+            method="Nelder-Mead",
+            options=options,
+        )
+
+        return process_scipy_result(res)
+```
+
+1. The new internal algorithms are dataclasses, where all algorithm options are
+   dataclass fields. This enables us to obtain information about the options via the
+   `__dataclass_fields__` attribute without inspecting signatures or imposing naming
+   conventions on non-option arguments.
+1. The `_solve_internal_problem` method receives an instance of `InternalProblem` and
+   `x0` (the start values) as arguments. `InternalProblem` collects the criterion
+   function, its derivatives, bounds, etc. This again avoids any potential for typos in
+   argument names.
+1. The `mark.minimizer` decorator collects all the information that was previously
+   collected via optional arguments with naming conventions. This information is
+   available while constructing the instance of `InternalProblem`. Thus we can make sure
+   that attributes that were not requested (e.g. derivatives if `needs_derivative` is
+   `False`) raise an `AttributeError` if used.
+1. The minimize function returns an `InternalOptimizeResult` instead of a dictionary.
+
+The copy constructors (`with_option`, `with_convergence`, and `with_stopping`) are
+inherited from `estimagic.Algorithm`. This means, that they will have `**kwargs` as
+signature and thus do not support autocomplete. However, they can check that all
+specified options are actually in the `__dataclass_fields__` and thus provide feedback
+before an optimization is run.
+
+All breaking changes of the internal algorithm interface are done without deprecation
+cycle.
+
+```{note}
+The `_solve_internal_problem` method is private because users should not call it; This
+also prepares adding a public `minimize` method that internally calls the
+`minimize` function.
+```
+
+To make things more concrete, here are prototypes for components related to the
+`InternalProblem` and `InternalOptimizeResult`.
+
+```python
+from numpy.typing import NDArray
+from dataclasses import dataclass
+from typing import Callable, Tuple
+import estimagic as em
+
+
+@dataclass(frozen=True)
+class ScalarProblemFunctions:
+    fun: Callable[[NDArray[float]], float]
+    jac: Callable[[NDArray[float]], NDArray[float]]
+    fun_and_jac: Callable[[NDArray[float]], Tuple[float, NDArray[float]]]
```

Comment:
This is the internal problem representation, where  everything is numpy based.  It is actually a goal to be as strict as possible here because then internal algorithms can make strong assumptions and have to handle fewer cases. 

The only reason to deviate from that would be if we want to support end-to-end jax compatible (and differentiable) optimizers like the ones from jaxopt (which is now part of optax). However that would require huge changes, as everything we do inside the internal criterion function would have to be jax compatible. 

I wouldn't say it's impossible but some features (e.g. logging) would be hard or impossible and of course the user's problems would also have to be fully jax compatible (e.g. no DataFrames as params).